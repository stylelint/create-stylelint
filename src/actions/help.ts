import { blue, dim } from 'picocolors';
import { log } from '../utils/logger.js';

export function showHelp(): void {
	const INDENT = {
		SECTION: ' '.repeat(2),
		OPTION: ' '.repeat(2),
	};

	const SPACES = {
		// Alignment for the second line:
		// Calculated as:
		// - "Usage: "          =  7 chars
		// - "create-stylelint" = 15 chars
		// - trailing space     =  2 spaces
		// Total                = 24 spaces
		SECOND_LINE: ' '.repeat(24),
	};

log(`Usage: create-stylelint [-v | --version] [-h | --help] [--dry-run] [--no-install | --no-color]
${SPACES.SECOND_LINE}[--use-npm | --use-pnpm | --use-yarn | --use-bun | --use-deno]

Options:
${INDENT.OPTION}--use-npm         ${dim('Use npm as package manager')}
${INDENT.OPTION}--use-pnpm        ${dim('Use pnpm as package manager')}
${INDENT.OPTION}--use-yarn        ${dim('Use yarn as package manager')}
${INDENT.OPTION}--use-bun         ${dim('Use bun as package manager')}
${INDENT.OPTION}--use-deno        ${dim('Use deno as package manager')}

${INDENT.OPTION}--dry-run         ${dim('Preview changes without applying them')}
${INDENT.OPTION}--no-install      ${dim('Skip dependency installation')}
${INDENT.OPTION}--no-color        ${dim('Disable color')}

${INDENT.OPTION}-h, --help        ${dim('Show this help message')}
${INDENT.OPTION}-v, --version     ${dim('Show version information')}

Examples:
${INDENT.OPTION}create-stylelint
${INDENT.OPTION}create-stylelint --use-npm --no-install

${dim('Need help?')} ${blue('https://github.com/stylelint/create-stylelint')}
`);
}
