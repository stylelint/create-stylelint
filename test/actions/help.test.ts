import { afterAll, afterEach, describe, expect, it, vi } from 'vitest';
import process from 'node:process';
import { showHelp } from '../../src/actions/help.js';

import { stripVTControlCharacters } from 'node:util';

describe('showHelp', () => {
	let output = '';
	const stdoutWriteSpy = vi.spyOn(process.stdout, 'write').mockImplementation((str) => {
		output += str;

		return true;
	});

	afterEach(() => {
		stdoutWriteSpy.mockClear();
		output = '';
	});

	afterAll(() => {
		stdoutWriteSpy.mockRestore();
	});

	it('should print help message to stdout', () => {
		showHelp();

		expect(stdoutWriteSpy).toHaveBeenCalledTimes(1);
		expect(output).toBeTruthy();
	});

	it('should include all package manager options', () => {
		showHelp();

		const cleaned = stripVTControlCharacters(output);

		const packageManagers = ['--use-npm', '--use-pnpm', '--use-yarn', '--use-bun', '--use-deno'];

		packageManagers.forEach((manager) => {
			expect(cleaned).toContain(manager);
		});
	});

	it('should include all general options', () => {
		showHelp();

		const cleaned = stripVTControlCharacters(output);

		const generalOptions = [
			'--dry-run',
			'--no-install',
			'--no-color',
			'-h, --help',
			'-v, --version',
		];

		generalOptions.forEach((option) => {
			expect(cleaned).toContain(option);
		});
	});

	it('should include usage examples', () => {
		showHelp();

		const cleaned = stripVTControlCharacters(output);

		expect(cleaned).toContain('Examples:');
		expect(cleaned).toContain('create-stylelint');
		expect(cleaned).toContain('create-stylelint --use-npm --no-install');
	});

	it('should include help link', () => {
		showHelp();

		const cleaned = stripVTControlCharacters(output);

		expect(cleaned).toContain('https://github.com/stylelint/create-stylelint');
	});

	it('should match snapshot', () => {
		showHelp();

		const cleaned = stripVTControlCharacters(output);

		expect(cleaned).toMatchSnapshot();
	});
});
