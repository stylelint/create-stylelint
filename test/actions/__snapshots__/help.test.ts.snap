// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`showHelp > should match snapshot 1`] = `
"Usage: create-stylelint [-v | --version] [-h | --help] [--dry-run] [--no-install | --no-color]
                        [--use-npm | --use-pnpm | --use-yarn | --use-bun | --use-deno]

Options:
  --use-npm         Use npm as package manager
  --use-pnpm        Use pnpm as package manager
  --use-yarn        Use yarn as package manager
  --use-bun         Use bun as package manager
  --use-deno        Use deno as package manager

  --dry-run         Preview changes without applying them
  --no-install      Skip dependency installation
  --no-color        Disable color

  -h, --help        Show this help message
  -v, --version     Show version information

Examples:
  create-stylelint
  create-stylelint --use-npm --no-install

Need help? https://github.com/stylelint/create-stylelint

"
`;

exports[`showHelp > should print full help message 1`] = `
"  create-stylelint [--version] [-h | --help] [--use-npm] [--use-pnpm]
    [--use-yarn] [--use-bun] [--dry-run] [--skip-install]
    [--no-color] [--no-input]

  Configuration options:
    -h, --help        Show this help message and exit
    --version         Display version information
    --no-color        Disable colored output
    --no-input        Disable interactive prompts

  Package manager selection:
    --use-npm         Use npm as package manager
    --use-pnpm        Use pnpm as package manager
    --use-yarn        Use yarn as package manager
    --use-bun         Use bun as package manager

  Installation options:
    --dry-run         Show what would be done without making changes
    --skip-install    Skip dependency installation step

  Examples:
    create-stylelint
    create-stylelint --use-pnpm --dry-run
    create-stylelint --no-input --use-npm

  For more information, visit:
    GitHub Repository: https://github.com/stylelint/create-stylelint
"
`;
